File: sdk/go/contract_test.go
// Copyright 2026 The Tormoni Authors. All rights reserved.
// Use of this source code is governed by the Apache-2.0 license that can be
// found in the LICENSE file.

package tormoni_test

// The two rules the whole contract rests on, and the ones most likely to be
// broken by a well-meaning change: a guest command exiting non-zero is a
// record and not an error, and an environment value goes out to the guest but
// never comes back in a record.

import (
	"context"
	"encoding/json"
	"reflect"
	"strings"
	"testing"

	tormoni "github.com/kendricklawton/tormoni/sdk/go"
)

// TestGuestExitIsNotAnError is the heart of the contract: the CLI exits with
// the guest command's own status, so a non-zero exit is a record to read, not
// an error to raise.
func TestGuestExitIsNotAnError(t *testing.T) {
	tests := []struct {
		name     string
		doc      string
		exitCode int
		wantKind tormoni.EndKind
		wantCode *int
		wantOK   bool
	}{
		{
			name:     "exited zero",
			doc:      `{"end_kind":"exit","end_code":0}`,
			exitCode: 0,
			wantKind: tormoni.EndExit, wantCode: ptr(0), wantOK: true,
		},
		{
			name:     "exited three",
			doc:      `{"end_kind":"exit","end_code":3}`,
			exitCode: 3,
			wantKind: tormoni.EndExit, wantCode: ptr(3), wantOK: false,
		},
		{
			name:     "exited one, the `false` case",
			doc:      `{"end_kind":"exit","end_code":1}`,
			exitCode: 1,
			wantKind: tormoni.EndExit, wantCode: ptr(1), wantOK: false,
		},
		{
			name:     "killed by a signal",
			doc:      `{"end_kind":"signal","end_code":9}`,
			exitCode: 137,
			wantKind: tormoni.EndSignal, wantCode: ptr(9), wantOK: false,
		},
		{
			name:     "stopped, with no code",
			doc:      `{"end_kind":"stopped","end_code":null}`,
			exitCode: 1,
			wantKind: tormoni.EndStopped, wantCode: nil, wantOK: false,
		},
		{
			name:     "gone, with no code",
			doc:      `{"end_kind":"gone","end_code":null}`,
			exitCode: 1,
			wantKind: tormoni.EndGone, wantCode: nil, wantOK: false,
		},
		{
			name:     "failed, with no code",
			doc:      `{"end_kind":"failed","end_code":null}`,
			exitCode: 1,
			wantKind: tormoni.EndFailed, wantCode: nil, wantOK: false,
		},
		{
			name:     "unknown, with no code",
			doc:      `{"end_kind":"unknown","end_code":null}`,
			exitCode: 1,
			wantKind: tormoni.EndUnknown, wantCode: nil, wantOK: false,
		},
		{
			name:     "still going",
			doc:      `{"end_kind":null,"end_code":null}`,
			exitCode: 0,
			wantKind: "", wantCode: nil, wantOK: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			s := newStub(t, tt.doc, "", tt.exitCode)
			run, err := s.Client().Run(context.Background(), []string{"sh", "-c", "whatever"}, nil)
			if err != nil {
				t.Fatalf("a guest exit of %d was raised as an error: %v", tt.exitCode, err)
			}
			if run.EndKind != tt.wantKind {
				t.Errorf("EndKind = %q, want %q", run.EndKind, tt.wantKind)
			}
			switch {
			case tt.wantCode == nil && run.EndCode != nil:
				t.Errorf("EndCode = %d, want nil", *run.EndCode)
			case tt.wantCode != nil && run.EndCode == nil:
				t.Errorf("EndCode = nil, want %d", *tt.wantCode)
			case tt.wantCode != nil && *run.EndCode != *tt.wantCode:
				t.Errorf("EndCode = %d, want %d", *run.EndCode, *tt.wantCode)
			}
			if run.OK() != tt.wantOK {
				t.Errorf("OK() = %v, want %v", run.OK(), tt.wantOK)
			}
		})
	}
}

// TestEnvValuesGoOutAndNamesComeBack pins the asymmetry: the guest gets the
// whole KEY=VALUE entry, and the record keeps only the name. No field of a Run
// carries a value, so nothing in this package can read one back.
func TestEnvValuesGoOutAndNamesComeBack(t *testing.T) {
	const secret = "s3cret-value-nobody-should-see"
	doc := `{"run_id":"1-run-1","end_kind":"exit","end_code":0,
	         "posture":{"env":["API_KEY","DEBUG"]}}`
	s := newStub(t, doc, "", 0)

	run, err := s.Client().Run(context.Background(), []string{"env"}, &tormoni.RunOptions{
		Env: []string{"API_KEY=" + secret, "DEBUG=1"},
	})
	if err != nil {
		t.Fatalf("run: %v", err)
	}

	// The value went out on the command line, whole.
	wantArgv := []string{"run", "--json",
		"--env", "API_KEY=" + secret,
		"--env", "DEBUG=1",
		"--", "env"}
	if got := s.Argv(); !reflect.DeepEqual(got, wantArgv) {
		t.Errorf("argv mismatch\n got: %q\nwant: %q", got, wantArgv)
	}

	// Only names came back.
	if !reflect.DeepEqual(run.Posture.Env, []string{"API_KEY", "DEBUG"}) {
		t.Errorf("Posture.Env = %q, want names only", run.Posture.Env)
	}

	// And no corner of the decoded record holds the value.
	encoded, err := json.Marshal(run)
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	if strings.Contains(string(encoded), secret) {
		t.Errorf("an environment value survived into the record: %s", encoded)
	}
}

File: sdk/js/tests/contract.test.ts
import { describe, expect, test } from "vitest";
import { Tormoni } from "../src/client";
import type { Run } from "../src/types";

// The two rules the whole contract rests on, and the ones most likely to be broken by a
// well-meaning change: an environment value goes out to the guest but never comes back in a
// record, and a guest command exiting non-zero is a record and not an error.
//
// These run against the REAL core, not a stub. There is no binary to stand in for any more: the
// SDK calls `execute_sandbox` in this process, so a fake would be testing itself. `dryRun`
// settles a posture without booting, which is what lets the contract be checked on a machine with
// no hypervisor.

const tormoni = new Tormoni();

describe("the environment asymmetry", () => {
  const SECRET = "s3cret-value-nobody-should-see";

  test("a value goes out whole and only the name comes back", async () => {
    const run = await tormoni.dryRun(["env"], { env: { API_KEY: SECRET, DEBUG: "1" } });
    expect(run.posture.env).toEqual(["API_KEY", "DEBUG"]);
  });

  test("no corner of the record holds the value", async () => {
    const run = await tormoni.dryRun(["env"], { env: { API_KEY: SECRET } });
    // Serialising the whole object is the point: a field added later that carried the value
    // would slip past an assertion naming only the fields that exist today.
    expect(JSON.stringify(run)).not.toContain(SECRET);
  });

  test("a value containing '=' survives the split", async () => {
    // `splitPair` once split on every `=`, which truncated a base64 value at its padding.
    const run = await tormoni.dryRun(["env"], { env: { TOKEN: "a=b=c" } });
    expect(run.posture.env).toEqual(["TOKEN"]);
  });
});

describe("the posture is the one that was asked for", () => {
  test("limits, network and rootfs come back as set", async () => {
    const run = await tormoni.dryRun(["true"], {
      vcpus: 2,
      mem: 1024,
      net: "tsi",
      rootfs: "writable",
    });
    expect(run.posture.vcpus).toBe(2);
    expect(run.posture.memMib).toBe(1024);
    expect(run.posture.network).toBe("tsi");
    expect(run.posture.rootfs).toBe("writable");
  });

  test("the defaults are the CLI's own, not this SDK's", async () => {
    const run = await tormoni.dryRun(["true"]);
    expect(run.posture.vcpus).toBe(1);
    expect(run.posture.memMib).toBe(512);
    expect(run.posture.network).toBe("none");
    expect(run.posture.rootfs).toBe("read-only");
    expect(run.posture.results).toBe(true);
  });

  test("mounts and shares keep their order and their pairing", async () => {
    const run = await tormoni.dryRun(["true"], {
      mounts: [["/mnt/a", "/tmp"], ["/mnt/b", "/var/tmp"]],
      shares: [["tag", "/tmp"]],
    });
    expect(run.posture.mounts).toEqual([["/mnt/a", "/tmp"], ["/mnt/b", "/var/tmp"]]);
    expect(run.posture.shares).toEqual([["tag", "/tmp"]]);
  });
});

describe("the boundary refuses what it cannot represent", () => {
  // `NonZeroU8` and `NonZeroU32` are the core's types. Zero has to be refused HERE, with a
  // sentence, rather than silently becoming the default on the way through.
  test("zero vcpus is refused, not rounded up", async () => {
    await expect(tormoni.dryRun(["true"], { vcpus: 0 })).rejects.toThrow(/at least 1|whole number/);
  });

  test("zero memory is refused", async () => {
    await expect(tormoni.dryRun(["true"], { mem: 0 })).rejects.toThrow(/at least 1/);
  });

  test("more vcpus than a u8 holds is refused", async () => {
    await expect(tormoni.dryRun(["true"], { vcpus: 256 })).rejects.toThrow(/1 to 255/);
  });

  test("an unknown posture word is refused rather than read as the default", async () => {
    // The dangerous version of this bug is silent: "writeable" falling through to read-only
    // gives a caller a sandbox they did not ask for and no sign of it.
    await expect(tormoni.dryRun(["true"], { rootfs: "writeable" as never })).rejects.toThrow(
      /read-only/,
    );
    await expect(tormoni.dryRun(["true"], { net: "host" as never })).rejects.toThrow(/none/);
  });

  test("an empty command is refused", async () => {
    await expect(tormoni.dryRun([])).rejects.toThrow(/empty/);
  });
});

describe("a dry run has settled a posture and nothing more", () => {
  test("it carries no end and no directory", async () => {
    const run: Run = await tormoni.dryRun(["echo", "hi"]);
    expect(run.command).toEqual(["echo", "hi"]);
    expect(run.verb).toBe("run");
    expect(run.endKind).toBeUndefined();
    expect(run.endedMs).toBeUndefined();
    expect(run.dir).toBeUndefined();
    expect(run.ok).toBe(false);
  });

  test("times are numbers, not BigInts", async () => {
    // napi maps `u64` to BigInt, which is contagious: `run.startedMs - t0` throws on a mix. The
    // Rust side uses i64 for exactly this reason, and this is what holds it there.
    const run = await tormoni.dryRun(["true"]);
    expect(typeof run.startedMs).toBe("number");
    expect(run.startedMs).toBeGreaterThan(0);
  });
});

describe("a missing run is a refusal a caller can read", () => {
  test("show names the id it could not find", async () => {
    await expect(tormoni.show("1-definitely-not-a-run")).rejects.toThrow(/no run/);
  });
});
